### PR TITLE
fib: allow for multiple instances of FIB

### PR DESCRIPTION
**Pull request by OlegHahm**
@authmillenon, @cgundogan, you might be interested, too.

IMO this is a) the cleaner way to go and b) required if we want to support dual stack applications (e.g. IPv6 and CCN or IPv6 and IPv4).

### CODE REVIEW

File: sys/net/network_layer/ng_ndp/ng_ndp.c [hunk 1]
```diff
@@ -34,6 +34,7 @@
 #include "net/ng_ndp.h"
 
 #define ENABLE_DEBUG    (0)
+
```
Comment (cgundogan) [inline sys/net/network_layer/ng_ndp/ng_ndp.c]: by mistake?

File: sys/include/net/fib.h [hunk 2]
```diff
@@ -169,29 +189,37 @@ int fib_get_next_hop(kernel_pid_t *iface_id,
...
+ *
+ * @param[in] table     the fib instance to check
  */
-int fib_get_num_used_entries(void);
+int fib_get_num_used_entries(fib_entry_t table[]);
 
 /**
  * @brief Prints the kernel_pid_t for all registered RRPs
+ *
+ * @param[in] table     the fib instance to print the entries from
```
Comment (cgundogan) [inline sys/include/net/fib.h]: you modified the doc here, but you did not adapt the code below.

**Comment by OlegHahm** [comment]
addressed comments

**Comment by BytesGalore** [comment]
Just a question for `b)` I don't get why this required to support a combinations of distinct endpoint/network numbering. Each entry in the FIB is unique through its destination address and this property doesn't change if you use distinct numberings.

**Comment by OlegHahm** [comment]
@BytesGalore, at least theoretically the identifiers for, e.g. CCN, and IPv6 may be syntactical identical, but with a completely different semantic, i.e. not unique. Assume, I have an entry for the IPv6 destination `beef::/64` in my FIB and perform a CCN lookup for `/beef/steak/yummy`. Another way around this, could to be either use the flags in the FIB or introduce a type field for the universal addresses.

**Comment by OlegHahm** [comment]
Rebased to master.

**Comment by cgundogan** [comment]
> /beef/steak/yummy

mmhh yummy beef steak

**Comment by BytesGalore** [comment]
@OlegHahm I see, that's true the FIB cannot handle collisions properly it would return the first "fitting" `next-hop`.
I guess all the FIB unittests [1] will fail, so they need to be adjusted too.

[1] https://github.com/RIOT-OS/RIOT/tree/master/tests/unittests/tests-fib

**Comment by cgundogan** [comment]
>Another way around this, could to be either use the flags in the FIB or introduce a type field for the universal addresses.

I prefer your current solution to have those addresses distributed across different tables. Otherwise, each lookup in the FIB would also require to check for the appropriate flag before determining the next-hop and that adds to the cost of the lookup.

**Comment by BytesGalore** [comment]
I'm with @cgundogan, I really like this PR :) its not just a cleaner way, we additionally get a pre-selection of searched address domains for free.

**Comment by OlegHahm** [comment]
On the downside this PR introduces 56 bytes more ROM without a real new feature. However, introduce an additional flag would cost RAM which is usually a bigger problem.

**Comment by BytesGalore** [comment]
That's true the advantage of this approach only applies if we really use multiple stacks/multiple addressing schemes at the same time.

> introduce an additional flag would cost RAM

also true, although the FIB entries have already flags, dispatching them would not come for free

**Comment by cgundogan** [comment]
@OlegHahm did you also think about giving a mutex for each table? Otherwise, an operation on one table would block the operation of another table. I am not sure though, if there are *long-lasting* operations in the fib @BytesGalore ? Maybe when talking to a routing protocol if no route is avail.?

**Comment by BytesGalore** [comment]
@cgundogan the search operations usually last longer, especially in the case for prefix matches since we need to check if we found the most fitting prefix and iterate all further entries.

I think the `universal_address` "pool" has more impact here, since its used across all FIB tables equally.
An individual mutex for each table would not help, since access to the universal addresses must be still mutually excluded.

**Comment by OlegHahm** [comment]
I would like to save the memory for per instance mutexes on the cost of potentially increased time for operations.

**Comment by BytesGalore** [comment]
+1 to keep it like it is

**Comment by OlegHahm** [comment]
So, all are happy?

**Comment by cgundogan** [comment]
I am happy, let's ask travis

**Comment by BytesGalore** [comment]
The unittests will fail until they are adopted to this change

**Comment by OlegHahm** [comment]
Ah, right, you mentioned this.

**Comment by OlegHahm** [comment]
Fixed. Should I squash?

**Comment by BytesGalore** [comment]
yes, please squash :)

**Comment by OlegHahm** [comment]
squashed

**Comment by BytesGalore** [comment]
ACK, when travis is happy

**Comment by OlegHahm** [comment]
Forgot to ifdef the FIB use in IPv6 header. Fixed and squashed immediately.

**Pull request merged**